File: dotnet/src/DocGen/Internal/Transport.cs
using System.Net;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using Dokmatiq.DocGen.Exceptions;

namespace Dokmatiq.DocGen.Internal;

/// <summary>
/// HTTP transport layer with retry logic and error mapping.
/// </summary>
internal sealed class Transport : IDisposable
{
    private static readonly HashSet<int> RetryableCodes = new() { 429, 500, 502, 503, 504 };

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
        PropertyNameCaseInsensitive = true,
        Converters = { new JsonStringEnumConverter() }
    };

    private readonly DocGenConfig _config;
    private readonly HttpClient _http;

    public Transport(DocGenConfig config)
    {
        _config = config;
        _http = new HttpClient { Timeout = config.Timeout };
    }

    /// <summary>Make a JSON request and deserialize the response.</summary>
    public async Task<T> RequestJsonAsync<T>(HttpMethod method, string path, object? body = null)
    {
        var bytes = await ExecuteAsync(method, path, SerializeBody(body), "application/json");
        return JsonSerializer.Deserialize<T>(bytes, JsonOptions)!;
    }

    /// <summary>Make a JSON request and return raw bytes.</summary>
    public async Task<byte[]> RequestBytesAsync(HttpMethod method, string path, object? body = null)
    {
        return await ExecuteAsync(method, path, SerializeBody(body), "application/json");
    }

    /// <summary>Make a JSON request and return a list.</summary>
    public async Task<List<T>> RequestListAsync<T>(HttpMethod method, string path)
    {
        var bytes = await ExecuteAsync(method, path, null, null);
        return JsonSerializer.Deserialize<List<T>>(bytes, JsonOptions) ?? new List<T>();
    }

    /// <summary>Upload a file via multipart/form-data and return deserialized response.</summary>
    public async Task<T> UploadAsync<T>(string path, string fieldName, byte[] fileBytes,
        string fileName, Dictionary<string, string>? extraFields = null)
    {
        var (body, contentType) = BuildMultipart(fieldName, fileBytes, fileName, extraFields);
        var bytes = await ExecuteAsync(HttpMethod.Post, path, body, contentType);
        return JsonSerializer.Deserialize<T>(bytes, JsonOptions)!;
    }

    /// <summary>Upload a file and return raw bytes.</summary>
    public async Task<byte[]> UploadBytesAsync(string path, string fieldName, byte[] fileBytes,
        string fileName, Dictionary<string, string>? extraFields = null)
    {
        var (body, contentType) = BuildMultipart(fieldName, fileBytes, fileName, extraFields);
        return await ExecuteAsync(HttpMethod.Post, path, body, contentType);
    }

    /// <summary>Send a DELETE request.</summary>
    public async Task DeleteAsync(string path)
    {
        await ExecuteAsync(HttpMethod.Delete, path, null, null);
    }

    // ── Synchronous wrappers ────────────────────────────────────────

    public T RequestJson<T>(HttpMethod method, string path, object? body = null)
        => RequestJsonAsync<T>(method, path, body).GetAwaiter().GetResult();

    public byte[] RequestBytes(HttpMethod method, string path, object? body = null)
        => RequestBytesAsync(method, path, body).GetAwaiter().GetResult();

    public List<T> RequestList<T>(HttpMethod method, string path)
        => RequestListAsync<T>(method, path).GetAwaiter().GetResult();

    public T Upload<T>(string path, string fieldName, byte[] fileBytes,
        string fileName, Dictionary<string, string>? extraFields = null)
        => UploadAsync<T>(path, fieldName, fileBytes, fileName, extraFields).GetAwaiter().GetResult();

    public byte[] UploadBytes(string path, string fieldName, byte[] fileBytes,
        string fileName, Dictionary<string, string>? extraFields = null)
        => UploadBytesAsync(path, fieldName, fileBytes, fileName, extraFields).GetAwaiter().GetResult();

    public void Delete(string path) => DeleteAsync(path).GetAwaiter().GetResult();

    // ── Private ─────────────────────────────────────────────────────

    private async Task<byte[]> ExecuteAsync(HttpMethod method, string path,
        byte[]? body, string? contentType)
    {
        var retry = _config.Retry;
        Exception? lastError = null;

        for (int attempt = 0; attempt <= retry.MaxRetries; attempt++)
        {
            try
            {
                using var request = new HttpRequestMessage(method, _config.BaseUrl + path);
                request.Headers.Add("X-API-Key", _config.ApiKey);

                if (_config.ValidateMode is not null)
                    request.Headers.Add("X-Validate-Mode", _config.ValidateMode);

                if (body is not null)
                {
                    request.Content = new ByteArrayContent(body);
                    if (contentType is not null)
                        request.Content.Headers.ContentType = MediaTypeHeaderValue.Parse(contentType);
                }

                using var response = await _http.SendAsync(request);
                var status = (int)response.StatusCode;

                if (status >= 200 && status < 300)
                    return await response.Content.ReadAsByteArrayAsync();

                if (RetryableCodes.Contains(status) && attempt < retry.MaxRetries)
                {
                    var delay = CalculateDelay(retry, attempt);

                    if (status == 429 && response.Headers.RetryAfter?.Delta is { } retryDelta)
                        delay = TimeSpan.FromMilliseconds(Math.Max(delay.TotalMilliseconds, retryDelta.TotalMilliseconds));

                    await Task.Delay(delay);
                    continue;
                }

                ThrowForStatus(status, response);
            }
            catch (ApiException) { throw; }
            catch (TaskCanceledException ex) when (ex.InnerException is TimeoutException)
            {
                throw new DocGenException("Request timed out", ex);
            }
            catch (HttpRequestException ex)
            {
                lastError = ex;
                if (attempt < retry.MaxRetries)
                {
                    await Task.Delay(CalculateDelay(retry, attempt));
                    continue;
                }
            }
        }

        throw new DocGenException("Request failed after all retries", lastError ?? new Exception("Unknown error"));
    }

    private static TimeSpan CalculateDelay(RetryPolicy retry, int attempt)
    {
        var delayMs = retry.InitialDelay.TotalMilliseconds * Math.Pow(retry.BackoffMultiplier, attempt);
        delayMs = Math.Min(delayMs, retry.MaxDelay.TotalMilliseconds);
        return TimeSpan.FromMilliseconds(delayMs);
    }

    private static void ThrowForStatus(int status, HttpResponseMessage response)
    {
        var body = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();
        string message;

        try
        {
            using var doc = JsonDocument.Parse(body);
            var root = doc.RootElement;
            message = root.TryGetProperty("message", out var msg) ? msg.GetString() ?? $"HTTP {status}"
                    : root.TryGetProperty("error", out var err) ? err.GetString() ?? $"HTTP {status}"
                    : $"HTTP {status}";
        }
        catch
        {
            message = $"HTTP {status}";
        }

        switch (status)
        {
            case 400:
                Dictionary<string, string>? fieldErrors = null;
                string? hint = null;
                try
                {
                    using var doc = JsonDocument.Parse(body);
                    if (doc.RootElement.TryGetProperty("fieldErrors", out var fe))
                        fieldErrors = JsonSerializer.Deserialize<Dictionary<string, string>>(fe.GetRawText());
                    if (doc.RootElement.TryGetProperty("hint", out var h))
                        hint = h.GetString();
                }
                catch { /* ignore parse errors */ }
                throw new ValidationException(message, body, fieldErrors, hint);

            case 401:
                throw new AuthenticationException(message);

            case 404:
                throw new NotFoundException(message);

            case 409:
                throw new ConflictException(message);

            case 429:
                double? retryAfter = null;
                int? limit = null, remaining = null;
                if (response.Headers.RetryAfter?.Delta is { } d)
                    retryAfter = d.TotalSeconds;
                if (response.Headers.TryGetValues("X-RateLimit-Limit", out var lv))
                    limit = int.TryParse(lv.FirstOrDefault(), out var li) ? li : null;
                if (response.Headers.TryGetValues("X-RateLimit-Remaining", out var rv))
                    remaining = int.TryParse(rv.FirstOrDefault(), out var ri) ? ri : null;
                throw new RateLimitException(message, retryAfter, limit, remaining);

            case 500:
                throw new ServerException(message, body);

            case 503:
                throw new ServiceUnavailableException(message, body);

            default:
                throw new ApiException(status, message, body);
        }
    }

    private static byte[]? SerializeBody(object? body)
    {
        if (body is null) return null;
        return JsonSerializer.SerializeToUtf8Bytes(body, JsonOptions);
    }

    private static (byte[] body, string contentType) BuildMultipart(
        string fieldName, byte[] fileBytes, string fileName, Dictionary<string, string>? extraFields)
    {
        var boundary = "----DocGenBoundary" + Guid.NewGuid().ToString("N");
        using var ms = new MemoryStream();
        using var writer = new StreamWriter(ms, Encoding.UTF8, leaveOpen: true) { NewLine = "\r\n" };

        writer.Write($"--{boundary}\r\n");
        writer.Write($"Content-Disposition: form-data; name=\"{fieldName}\"; filename=\"{fileName}\"\r\n");
        writer.Write("Content-Type: application/octet-stream\r\n\r\n");
        writer.Flush();
        ms.Write(fileBytes, 0, fileBytes.Length);

        if (extraFields is not null)
        {
            foreach (var (key, value) in extraFields)
            {
                writer.Write($"\r\n--{boundary}\r\n");
                writer.Write($"Content-Disposition: form-data; name=\"{key}\"\r\n\r\n");
                writer.Write(value);
            }
        }

        writer.Write($"\r\n--{boundary}--\r\n");
        writer.Flush();

        return (ms.ToArray(), $"multipart/form-data; boundary={boundary}");
    }

    public void Dispose() => _http.Dispose();
}

File: typescript/src/transport.ts
import type { RetryPolicy } from "./config.js";
import {
  ApiError,
  AuthenticationError,
  ConflictError,
  NotFoundError,
  RateLimitError,
  ServerError,
  ServiceUnavailableError,
  TimeoutError,
  ValidationError,
} from "./errors.js";

const RETRYABLE_STATUS_CODES = new Set([429, 500, 502, 503, 504]);

interface TransportConfig {
  baseUrl: string;
  apiKey: string;
  timeout: number;
  retry: Required<RetryPolicy>;
  validateMode?: "strict" | "warn";
}

/** HTTP transport layer with retry and error mapping. */
export class Transport {
  private readonly config: TransportConfig;

  constructor(config: TransportConfig) {
    this.config = config;
  }

  /** Make a JSON API request and return parsed response. */
  async requestJson<T>(
    method: string,
    path: string,
    body?: unknown,
  ): Promise<T> {
    const response = await this.request(method, path, {
      headers: { "Content-Type": "application/json" },
      body: body != null ? JSON.stringify(body) : undefined,
    });
    return (await response.json()) as T;
  }

  /** Make an API request and return raw bytes. */
  async requestBytes(
    method: string,
    path: string,
    body?: unknown,
  ): Promise<Buffer> {
    const response = await this.request(method, path, {
      headers: { "Content-Type": "application/json" },
      body: body != null ? JSON.stringify(body) : undefined,
    });
    const arrayBuffer = await response.arrayBuffer();
    return Buffer.from(arrayBuffer);
  }

  /** Upload a file via multipart form data. */
  async upload<T>(
    path: string,
    fieldName: string,
    fileBuffer: Buffer,
    fileName: string,
    extraFields?: Record<string, string>,
  ): Promise<T> {
    const formData = new FormData();
    formData.append(
      fieldName,
      new Blob([new Uint8Array(fileBuffer)]),
      fileName,
    );
    if (extraFields) {
      for (const [key, value] of Object.entries(extraFields)) {
        formData.append(key, value);
      }
    }
    const response = await this.request("POST", path, { body: formData });
    return (await response.json()) as T;
  }

  /** Upload raw bytes and return binary response. */
  async uploadBytes(
    path: string,
    fieldName: string,
    fileBuffer: Buffer,
    fileName: string,
    extraFields?: Record<string, string>,
  ): Promise<Buffer> {
    const formData = new FormData();
    formData.append(
      fieldName,
      new Blob([new Uint8Array(fileBuffer)]),
      fileName,
    );
    if (extraFields) {
      for (const [key, value] of Object.entries(extraFields)) {
        formData.append(key, value);
      }
    }
    const response = await this.request("POST", path, { body: formData });
    const arrayBuffer = await response.arrayBuffer();
    return Buffer.from(arrayBuffer);
  }

  /** Send a DELETE request. */
  async delete(path: string): Promise<void> {
    await this.request("DELETE", path);
  }

  /** Make a JSON request and return an array. */
  async requestList<T>(method: string, path: string): Promise<T[]> {
    const response = await this.request(method, path);
    return (await response.json()) as T[];
  }

  private async request(
    method: string,
    path: string,
    options?: { headers?: Record<string, string>; body?: string | FormData },
  ): Promise<Response> {
    const url = `${this.config.baseUrl}${path}`;
    const headers: Record<string, string> = {
      "X-API-Key": this.config.apiKey,
      ...options?.headers,
    };
    if (this.config.validateMode) {
      headers["X-Validate-Mode"] = this.config.validateMode;
    }

    const { maxRetries, initialDelay, backoffMultiplier, maxDelay } =
      this.config.retry;
    let lastError: Error | undefined;

    for (let attempt = 0; attempt <= maxRetries; attempt++) {
      try {
        const controller = new AbortController();
        const timeoutId = setTimeout(
          () => controller.abort(),
          this.config.timeout,
        );

        const response = await fetch(url, {
          method,
          headers,
          body: options?.body,
          signal: controller.signal,
        });

        clearTimeout(timeoutId);

        if (response.ok) {
          return response;
        }

        // Check if retryable
        if (
          RETRYABLE_STATUS_CODES.has(response.status) &&
          attempt < maxRetries
        ) {
          let delay = initialDelay * Math.pow(backoffMultiplier, attempt);
          delay = Math.min(delay, maxDelay);

          // Respect Retry-After header on 429
          if (response.status === 429) {
            const retryAfter = response.headers.get("Retry-After");
            if (retryAfter) {
              const retryMs = parseFloat(retryAfter) * 1000;
              if (!isNaN(retryMs)) {
                delay = Math.max(delay, retryMs);
              }
            }
          }

          await sleep(delay);
          continue;
        }

        // Non-retryable or retries exhausted — throw mapped error
        await throwForStatus(response);
      } catch (error: unknown) {
        if (error instanceof ApiError) {
          throw error;
        }
        if (
          error instanceof Error &&
          error.name === "AbortError"
        ) {
          lastError = new TimeoutError(
            `Request timed out after ${this.config.timeout}ms`,
          );
          if (attempt < maxRetries) {
            const delay = Math.min(
              initialDelay * Math.pow(backoffMultiplier, attempt),
              maxDelay,
            );
            await sleep(delay);
            continue;
          }
          throw lastError;
        }
        lastError = error instanceof Error ? error : new Error(String(error));
        if (attempt < maxRetries) {
          const delay = Math.min(
            initialDelay * Math.pow(backoffMultiplier, attempt),
            maxDelay,
          );
          await sleep(delay);
          continue;
        }
      }
    }

    throw lastError ?? new Error("Request failed after all retries");
  }
}

async function throwForStatus(response: Response): Promise<never> {
  const body = await response.text();
  let parsed: Record<string, unknown> | undefined;
  try {
    parsed = JSON.parse(body);
  } catch {
    // not JSON
  }

  const message =
    (parsed?.["message"] as string) ??
    (parsed?.["error"] as string) ??
    `HTTP ${response.status}`;

  switch (response.status) {
    case 400: {
      const fieldErrors = parsed?.["fieldErrors"] as
        | Record<string, string>
        | undefined;
      const hint = parsed?.["hint"] as string | undefined;
      throw new ValidationError(message, body, fieldErrors, hint);
    }
    case 401:
      throw new AuthenticationError(message);
    case 404:
      throw new NotFoundError(message);
    case 409:
      throw new ConflictError(message);
    case 429: {
      const retryAfter = response.headers.get("Retry-After");
      const limit = response.headers.get("X-RateLimit-Limit");
      const remaining = response.headers.get("X-RateLimit-Remaining");
      throw new RateLimitError(
        message,
        retryAfter ? parseFloat(retryAfter) : undefined,
        limit ? parseInt(limit, 10) : undefined,
        remaining ? parseInt(remaining, 10) : undefined,
      );
    }
    case 500:
      throw new ServerError(message, body);
    case 503:
      throw new ServiceUnavailableError(message, body);
    default:
      throw new ApiError(response.status, message, body);
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
